ns1/resource_region.go

```diff
@@ -0,0 +1,267 @@
+	m := make(map[string]interface{})
+	name, region := getRegion(regions)
+	m["name"] = name
+	if region != nil {
+		m["meta"] = region.Meta.StringMap()
+	}
+	return m
+}
+
+func regionsToResourceData(d *schema.ResourceData, regions data.Regions) error {
+	m := regionToMap(regions)
+	d.Set("name", m["name"])
+	if m["meta"] != nil {
+		d.Set("meta", m["meta"])
+	}
+	d.SetId(regionIDHash(d))
+	return nil
+}
+
+func resourceDataToRegions(regions data.Regions, d *schema.ResourceData, old bool) error {
+	var name string
+	var meta map[string]interface{}
+	oldName, newName := d.GetChange("name")
+	oldMeta, newMeta := d.GetChange("meta")
+	if old {
+		name = oldName.(string)
+		meta = oldMeta.(map[string]interface{})
+	} else {
+		name = newName.(string)
+		meta = newMeta.(map[string]interface{})
+	}
+	var region data.Region
+	if meta != nil {
+		region.Meta = *data.MetaFromMap(meta)
+		errs := region.Meta.Validate()
+		if len(errs) > 0 {
+			return errJoin(append([]error{errors.New("found error/s in region metadata")}, errs...), ",")
+		}
+	}
+	regions[name] = region
+	return nil
+}
+
+func findRecordByRegion(client *ns1.Client, domain string) (*dns.Record, error) {
+	zones, _, err := client.Zones.List()
+	if err != nil {
+		return nil, err
+	}
+	for _, z := range zones {
+		zone, _, err := client.Zones.Get(z.Zone)
+		if err != nil {
+			return nil, err
+		}
+		for _, record := range zone.Records {
+			if domain == record.Domain {
+				r, _, err := client.Records.Get(z.Zone, record.Domain, record.Type)
+				if err != nil {
+					return nil, err
+				}
+				return r, err
+			}
+		}
+	}
+	return nil, fmt.Errorf("record not found: %s", domain)
+}
+
+func findRegion(d *schema.ResourceData, record *dns.Record, old bool) (data.Regions, error) {
+	var regions = data.Regions{}
+	if err := resourceDataToRegions(regions, d, true); err != nil {
+		return nil, err
+	}
+	name, _ := getRegion(regions)
+	for k, v := range record.Regions {
+		if k != name {
+			continue
+		}
+		regions[k] = v
+		return regions, nil
+	}
+	return nil, nil
+}
+
+func updateRecordForRegion(op string, meta interface{}, d *schema.ResourceData) (data.Regions, error) {
+	client := meta.(*ns1.Client)
+	var regions = data.Regions{}
+	// get the record to get the zone before creating lock
+	r, err := findRecordByRegion(client, d.Get("record").(string))
+	if err != nil {
+		return nil, err
+	}
+	err = RecordMutex.Lock(client, &regions, d.Get("record").(string), r.Zone)
+	if err != nil {
+		return nil, err
+	}
+	defer RecordMutex.Unlock(client, &regions, d.Get("record").(string), r.Zone)
+	// get the record again after creating lock
+	r, err = findRecordByRegion(client, d.Get("record").(string))
+	if err != nil {
+		return nil, err
+	}
+	if err := resourceDataToRegions(regions, d, false); err != nil {
+		return nil, err
+	}
+	switch op {
+	case "create":
+		// Create the region
+		region := data.Region{
+			Meta: data.Meta{},
+		}
+		meta := data.MetaFromMap(d.Get("meta").(map[string]interface{}))
```

Comment:
MetaFromMap returns a pointer, it seems, so we should check that it's not nil before dereferencing it.

Reply:
(also in `update` below)

Reply:
i hope i did it correctly (in commit de11d8f)